runtime/src/accounts_db.rs

```diff
@@ -1801,6 +2309,96 @@ impl AccountsDB {
+            if let Some((_, slot_removed_storages)) = self.storage.0.remove(&slot) {
+                {
+                    let r_slot_removed_storages = slot_removed_storages.read().unwrap();
+                    total_removed_storage_entries += r_slot_removed_storages.len();
+                    total_removed_bytes += r_slot_removed_storages
+                        .values()
+                        .map(|i| i.accounts.capacity())
+                        .sum::<u64>();
+                }
+                all_removed_slot_storages.push(slot_removed_storages.clone());
+            }
+        }
+        remove_storages_elapsed.stop();
+
+        let num_slots_removed = all_removed_slot_storages.len();
+
+        let recycle_stores_write_time =
+            self.recycle_slot_stores(total_removed_storage_entries, &all_removed_slot_storages);
+
+        let mut drop_storage_entries_elapsed = Measure::start("drop_storage_entries_elapsed");
+        // Backing mmaps for removed storages entries explicitly dropped here outside
+        // of any locks
+        drop(all_removed_slot_storages);
+        drop_storage_entries_elapsed.stop();
+
+        datapoint_info!(
+            "purge_slots_time",
+            ("safety_checks_elapsed", safety_checks_elapsed.as_us(), i64),
+            (
+                "remove_storages_elapsed",
+                remove_storages_elapsed.as_us(),
+                i64
+            ),
+            (
+                "drop_storage_entries_elapsed",
+                drop_storage_entries_elapsed.as_us(),
+                i64
+            ),
+            ("num_slots_removed", num_slots_removed, i64),
+            (
+                "total_removed_storage_entries",
+                total_removed_storage_entries,
+                i64
+            ),
+            ("total_removed_bytes", total_removed_bytes, i64),
+            (
+                "recycle_stores_write_elapsed",
+                recycle_stores_write_time,
+                i64
+            ),
+        );
+    }
+
+    fn purge_slot_cache_keys(&self, slot: Slot, slot_cache: SlotCache) -> Vec<(u64, AccountInfo)> {
+        // Slot purged from cache should not exist in the backing store
+        assert!(self.storage.get_slot_stores(slot).is_none());
+        let dead_slot: HashSet<Slot> = vec![slot].into_iter().collect();
+        let mut purged_slot_pubkeys: HashSet<(Slot, Pubkey)> = HashSet::new();
+        let pubkey_to_slot_set: Vec<(Pubkey, HashSet<Slot>)> = slot_cache
+            .iter()
+            .map(|account| {
+                purged_slot_pubkeys.insert((slot, *account.key()));
+                (*account.key(), dead_slot.clone())
+            })
+            .collect();
+        let num_purged_keys = pubkey_to_slot_set.len();
+        let reclaims = self.purge_keys_exact(pubkey_to_slot_set);
+        assert_eq!(reclaims.len(), num_purged_keys);
+        self.clean_dead_slots_common(&dead_slot, purged_slot_pubkeys, None);
+        reclaims
```

Comment:
also, it looks like this returned value isn't used by caller at all?

Reply:
good, catch removed